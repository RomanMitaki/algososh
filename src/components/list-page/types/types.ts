class Node<T> {
    value: T;
    next: Node<T> | null;

    constructor(value: T, next?: Node<T> | null) {
        this.value = value;
        this.next = next === undefined ? null : next;
    }
}

interface ILinkedList<T> {
    prepend: (element: T) => void;
    append: (element: T) => void;
    deleteHead: () => void;
    deleteTail: () => void;
    addByIndex: (element: T, index: number) => void;
    deleteByIndex: (index: number) => void;
    toArray: () => T[] | null;
}

export class LinkedList<T> implements ILinkedList<T> {
    private head: Node<T> | null;
    private tail: Node<T> | null;
    private size: number;

    constructor(array?: T[]) {
        this.head = null;
        this.size = 0;
        this.tail = null;
        array?.forEach(element => this.append(element));
    }


    prepend(element: T) {
        const node = new Node(element);
        if (!this.head) {
            this.head = node;
            this.tail = node;
        } else {
            node.next = this.head;
            this.head = node;
        }
        this.size++;
    }

    deleteHead() {
        if (this.size === 1) {
            this.head = null;
            this.size = 0;
            this.tail = null;
        } else {
            this.head = this.head!.next;
            this.size--;
        }
    }

    append(element: T) {
        const node = new Node(element);
        if (!this.tail) {
            this.tail = node;
            this.head = node;
        } else if (this.size === 1) {
            this.head!.next = node;
            this.tail = node;
        } else {
            let n = this.head!.next;
            while (this.size > 1 && n?.next) {
                n = n.next;
            }
            n!.next = node;
            this.tail = node;
        }
        this.size++;
    }

    deleteTail() {
        if (this.size === 1) {
            this.head = null;
            this.size = 0;
            this.tail = null;
        } else {
            let n = this.head;
            let prev = this.head;
            while (this.size > 1 && n?.next) {
                n = n.next;
                prev = prev!.next;
            }
            prev!.next = null;
            this.tail = prev;
            this.size--;
        }
    }

    addByIndex(element: T, index: number) {
        if (index < 0 || index > this.size) {
            console.log("Enter a valid index");
            return;
        } else {
            const node = new Node(element);
            // добавить элемент в начало списка
            if (index === 0) {
                node.next = this.head;
                this.head = node;
            } else {
                let curr = this.head;
                // перебрать элементы в списке до нужной позиции
                while (index > 1 && curr) {
                    curr = curr.next;
                    index--;
                }
                // добавить элемент
                if (curr) {
                    let next = curr.next;
                    curr.next = node;
                    if (next) {
                        node.next = next;
                    } else {
                        this.tail = curr.next;
                    }
                }
            }
            this.size++;
        }
    }

    deleteByIndex(index: number) {
        if (index < 0 || index > this.size || !this.head) {
            return;
        }
        if (this.size === 1) {
            this.head = null;
            this.size = 0;
            this.tail = null;
        }
        let curr = this.head;
        if (index === 0) {
            this.head = curr!.next;
            curr!.next = null;
        } else {
            let prev = null;
            let n = 0;
            while (n < index) {
                prev = curr;
                if (curr) curr = curr?.next;
                n++;
            }
            if (prev && curr) prev.next = curr.next ? curr.next : null;
            if (curr) curr.next = null;
        }
        this.size--;
    }

    /*getSize() {
        return this.size;
    }*/

    toArray(): T[] | null {
        let curr = this.head;
        let res = [];
        while (curr) {
            res.push(curr.value);
            curr = curr.next;
        }
        return res;
    }


}
