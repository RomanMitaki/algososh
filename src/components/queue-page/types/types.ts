interface IQueue<T> {
    enqueue: (item: T) => void;
    dequeue: () => void;
    peak: () => T | null;
    print: () => (T | null)[];
}

export class Queue<T> implements IQueue<T> {
    private container: (T | null)[] = [];
    private head = 0;
    private tail = 0;
    private readonly size: number = 0;
    private length: number = 0;

    constructor(size: number) {
        this.size = size;
        this.container = Array(size).fill('');
    }

    enqueue = (item: T) => {
        if (this.length >= this.size) {
            throw new Error("Maximum length exceeded");
        }
        this.container[this.tail % this.size] = item;
        this.length++;
        if (this.tail < this.size - 1) {
            this.tail++;
        } else {
            this.tail = 0;
        }
    };

    dequeue = () => {
        if (this.isEmpty()) {
            throw new Error("No elements in the queue");
        }
        delete this.container[this.head % this.size];
        this.length--;
        if (this.head < this.size - 1) {
            this.head++;
        } else {
            this.head = 0;
        }
    };

    peak = (): T | null => {
        if (this.isEmpty()) {
            throw new Error("No elements in the queue");
        }
        return this.container[this.head % this.size]; // Ваш код
    };

    isEmpty = () => this.length === 0;

    print = (): (T | null)[] => {
        return this.container.map((element) => {
            return element
        })
    };
}