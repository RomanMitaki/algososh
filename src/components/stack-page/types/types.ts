interface IStack<T> {
    push: (item: T) => void;
    pop: () => void;
    peak: () => T | null;
    getSize: () => number;
    print: () => T[] | null;
}

export class Stack<T> implements IStack<T> {
    private container: T[] = [];

    push = (item: T): void => {
        this.container.push(item);
    };

    pop = (): void => {
        if (this.container) {
            this.container.pop();
        }
    };

    peak = (): T | null => {
        if (this.container) {
            return this.container[this.container.length - 1]
        }
        return null;
    };

    print = (): T[] | null => {
        if (this.container) {
            return this.container.map((element, idx) => {
                return element;
            });
        }
        return null;
    };

    getSize = () => this.container.length;
}