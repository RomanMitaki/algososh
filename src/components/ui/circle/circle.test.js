import {Circle} from "./circle";
import renderer from "react-test-renderer";
import {ElementStates} from "../../../types/element-states";

describe('Тест Circle', () => {
    it('Рендер без буквы', () => {
        const circle = renderer.create(<Circle/>).toJSON();
        expect(circle).toMatchSnapshot()
    })

    it('Рендер с буквами', () => {
        const circle = renderer.create(<Circle letter={'test'}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с head text', () => {
        const circle = renderer.create(<Circle head={'test'}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с head null', () => {
        const circle = renderer.create(<Circle head={null}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с head ReactElement', () => {
        const circle = renderer.create(<Circle head={<Circle/>}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с tail text', () => {
        const circle = renderer.create(<Circle tail="test"/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с tail ReactElement', () => {
        const circle = renderer.create(<Circle tail={<Circle/>}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер с index', () => {
        const circle = renderer.create(<Circle index="0"/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер isSmall', () => {
        const circle = renderer.create(<Circle isSmall={true}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер в состоянии ElementStates.default', () => {
        const circle = renderer.create(<Circle state={ElementStates.Default}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер в состоянии ElementStates.changing', () => {
        const circle = renderer.create(<Circle state={ElementStates.Changing}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

    it('Рендер в состоянии ElementStates.modified', () => {
        const circle = renderer.create(<Circle state={ElementStates.Modified}/>).toJSON();
        expect(circle).toMatchSnapshot();
    })

})