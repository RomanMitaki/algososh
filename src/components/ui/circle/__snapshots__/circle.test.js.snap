// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Тест Circle Рендер isSmall 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle  small default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер без буквы 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер в состоянии ElementStates.changing 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   changing"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер в состоянии ElementStates.default 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер в состоянии ElementStates.modified 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   modified"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с head 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    test
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с head 2`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с head ReactElement 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с head null 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с head text 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head string"
  >
    test
  </div>
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;

exports[`Тест Circle Рендер с index 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  >
    0
  </p>
  <div
    className="text text_type_input text_color_input mt-4 absolute tail60 element"
  />
</div>
`;

exports[`Тест Circle Рендер с tail ReactElement 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  >
    <div
      className="content "
    >
      <div
        className="text text_type_input text_color_input mb-4 absolute head element"
      />
      <div
        className="circle   default"
      >
        <p
          className="text text_type_circle text_color_input letter"
        />
      </div>
      <p
        className="text text_type_input text_color_input mt-4 absolute index"
      />
      <div
        className="text text_type_input text_color_input mt-4 absolute tail30 element"
      />
    </div>
  </div>
</div>
`;

exports[`Тест Circle Рендер с tail text 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    />
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 string"
  >
    test
  </div>
</div>
`;

exports[`Тест Circle Рендер с буквами 1`] = `
<div
  className="content "
>
  <div
    className="text text_type_input text_color_input mb-4 absolute head element"
  />
  <div
    className="circle   default"
  >
    <p
      className="text text_type_circle text_color_input letter"
    >
      test
    </p>
  </div>
  <p
    className="text text_type_input text_color_input mt-4 absolute index"
  />
  <div
    className="text text_type_input text_color_input mt-4 absolute tail30 element"
  />
</div>
`;
